Change JetStream benchmark harness to pre-fetch files first.

MenloDorian opened the pull request:
#### 5186d5c427b58d37a83770006221c541e5415690
<pre>
Change JetStream benchmark harness to pre-fetch files first.
<a href="https://bugs.webkit.org/show_bug.cgi?id=244412">https://bugs.webkit.org/show_bug.cgi?id=244412</a>
&lt;rdar://problem/98787675&gt;

Reviewed by Yusuke Suzuki.

This improves the stability of the benchmark score by reducing the impact of
network latency and consequent any second order effects.

* PerformanceTests/JetStream2/JetStream.css
* PerformanceTests/JetStream2/JetStreamDriver.js:
(Driver.prototype.async initialize):
(Driver.prototype.async preloadFilesForBrowser):
(prototype.async run):
(prototype.async loadBlob):
(prototype.preloadFilesForBrowser):
(prototype.fetchResources)
* PerformanceTests/JetStream2/in-depth.html:
* PerformanceTests/JetStream2/index.html:

Canonical link: <a href="https://commits.webkit.org/254033@main">https://commits.webkit.org/254033@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/dbe1752c4b3698a7e9df8ca547c92f3bc36a4daa

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/6/builds/87858 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/69/builds/31952 "Built successfully") | [  ~~🛠 mac~~](https://ews-build.webkit.org/#/builders/43/builds/18570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/8/builds/97027 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/64/builds/30313 "Built successfully") | [  ~~🛠 mac-debug~~](https://ews-build.webkit.org/#/builders/71/builds/26362 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/36/builds/79944 "Built successfully") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/12/builds/91771 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/19/builds/93472 "Passed tests") | [  ~~🧪 ios-wk2~~](https://ews-build.webkit.org/#/builders/68/builds/24473 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/61/builds/74569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/35/builds/24446 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/9/builds/79430 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/43/builds/18570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/34/builds/67291 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/66/builds/27978 "Built successfully") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/43/builds/18570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/67/builds/27962 "Built successfully") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/43/builds/18570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/63/builds/29655 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/61/builds/74569 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/75/builds/1153 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/65/builds/29543 "Built successfully") | [  ~~🧪 mac-wk2-stress~~](https://ews-build.webkit.org/#/builders/43/builds/18570 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
<!--EWS-Status-Bubble-End-->

saambarati commented on PerformanceTests/JetStream2/in-depth.html:
```diff
@@ -113,12 +113,20 @@ <h2>In-Depth Analysis</h2>
+        improve score stability. This is achieved by pre-fetching network resources to warm up the
+        browser network cache prior to running the benchmarks. This reduces perturbations on the
+        measurement of JavaScript execution time due to network latency and any consequent second
+        order effects.
```
maybe instead of "this reduces" we could say "This can reduce" or "This may reduce"

MenloDorian replied:
Fixed.

Constellation commented on PerformanceTests/JetStream2/JetStreamDriver.js:
```diff
@@ -597,6 +609,19 @@ class Benchmark {
+                preloads.push(this.plan.preload[prop]);
+        }
+        preloads = preloads.map((file) => fileLoader.load(file));
+        return Promise.all([filePromises, preloads]);
```
This is passing array of array of promises, so this function will not wait for all promises' resolution.

MenloDorian replied:
Fixed.

Constellation commented on PerformanceTests/JetStream2/JetStreamDriver.js:
```diff
@@ -597,11 +636,71 @@ class Benchmark {
+        if (blobData) {
+            blobData[3]++;
+            return blobData;
+        }
```
How about chaining a promise of blob in the cache instead?
Then we do not need to request the same file multiple times.
```

let promise = cache[file];
if (promise)
    return await promise;
...
promise = new Promise((resolve) => {
    let tries = 3;
    ...

    When fetch resource is resolved, call resolve(blob);
});
cache[file] = promise;
```

MenloDorian replied:
I can try it.

Constellation reviewed (commented):
Looks nice. I have some comments.

Constellation commented on PerformanceTests/JetStream2/JetStreamDriver.js:
```diff
@@ -233,6 +244,16 @@ class Driver {
             benchmark.updateUIAfterRun();
+
+            if (isInBrowser) {
+                let cache = window.blobDataCache;
```
Let's make `this.blobDataCache`. since `Driver` is alive, we do not need to create a window's variable.

Constellation replied:
And let's have `this.blobDataCache = {};` in the constructor of this `Driver`.

MenloDorian replied:
I'll put it in the `Driver` but reference it using `JetStream.blobDataCache`.  `JetStream` is the global singleton instance of `Driver`.  Some of the places below are in the `Benchmark` class and not `Driver` itself.  I'll just use `JetStream.blobDataCache` everywhere for consistency.

Ditto for `JetStream.loadCache`.

Constellation commented on PerformanceTests/JetStream2/JetStreamDriver.js:
```diff
@@ -371,16 +392,35 @@ class Driver {
+            promises.push(benchmark.preloadFilesForBrowser(counter));
+        await Promise.all(promises);
+
+        window.loadCache = undefined; // Done preloading all the files.
```
Let's make it `this.loadCache`. And let's clear it with `this.loadCache = {};`. So we always have an object.
And let's put `this.loadCache = {};` in the constructor of this `Driver`.

Constellation commented on PerformanceTests/JetStream2/JetStreamDriver.js:
```diff
@@ -545,14 +585,15 @@ class Benchmark {
             for (let text of this.scripts)
                 addScript(text);
         } else {
+            let cache = window.blobDataCache;
```
Let's make it `this.blobDataCache`.

Constellation commented on PerformanceTests/JetStream2/JetStreamDriver.js:
```diff
@@ -597,11 +638,67 @@ class Benchmark {
+        }
+        let blob = await response.blob();
+        var blobData = [ file, blob, URL.createObjectURL(blob), 0 ];
+        window.blobDataCache[file] = blobData;
```
Ditto.

Constellation commented on PerformanceTests/JetStream2/JetStreamDriver.js:
```diff
@@ -597,11 +638,67 @@ class Benchmark {
+        if (!window.blobDataCache) {
+            window.blobDataCache = { };
+            window.loadCache = { };
+        }
```
I think we don't need it if we initialize them in Driver's constructor.

Constellation commented on PerformanceTests/JetStream2/JetStreamDriver.js:
```diff
@@ -597,11 +638,67 @@ class Benchmark {
+            window.loadCache = { };
+        }
+
+        let promise = window.loadCache[file];
```
this.loadCache.

Constellation commented on PerformanceTests/JetStream2/JetStreamDriver.js:
```diff
@@ -597,11 +638,67 @@ class Benchmark {
+            return await promise;
+
+        promise = this.doLoadBlob(file);
+        window.loadCache[file] = promise;
```
Ditto.

Constellation reviewed (approved):
r=me

Commit:
Change JetStream benchmark harness to pre-fetch files first.
https://bugs.webkit.org/show_bug.cgi?id=244412
<rdar://problem/98787675>

Reviewed by Yusuke Suzuki.

This improves the stability of the benchmark score by reducing the impact of
network latency and consequent any second order effects.

* PerformanceTests/JetStream2/JetStream.css
* PerformanceTests/JetStream2/JetStreamDriver.js:
(Driver.prototype.async initialize):
(Driver.prototype.async preloadFilesForBrowser):
(prototype.async run):
(prototype.async loadBlob):
(prototype.preloadFilesForBrowser):
(prototype.fetchResources)
* PerformanceTests/JetStream2/in-depth.html:
* PerformanceTests/JetStream2/index.html:

Canonical link: https://commits.webkit.org/254033@main
